File: src/main/java/de/ayont/lpc/listener/SpigotChatListener.java
package de.ayont.lpc.listener;

import de.ayont.lpc.LPC;
import net.kyori.adventure.text.Component;
import de.ayont.lpc.renderer.SpigotChatRenderer;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import org.bukkit.Material;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.HashMap;
import java.util.Map;


public class SpigotChatListener implements Listener {
    private final LPC plugin;
    private final SpigotChatRenderer chatRenderer;
    private final Map<String, String> legacyToMiniMessageColors;

    public SpigotChatListener(LPC plugin) {
        this.plugin = plugin;
        this.chatRenderer = new SpigotChatRenderer(plugin);
        this.legacyToMiniMessageColors = new HashMap<>();
        initColorMappings();
    }

    private String formatTimeLeft(long millis) {
        if (millis <= 0) return "0s";
        long seconds = millis / 1000 % 60;
        long minutes = millis / (60 * 1000) % 60;
        long hours = millis / (60 * 60 * 1000) % 24;
        long days = millis / (24 * 60 * 60 * 1000);

        StringBuilder sb = new StringBuilder();
        if (days > 0) sb.append(days).append("d ");
        if (hours > 0) sb.append(hours).append("h ");
        if (minutes > 0) sb.append(minutes).append("m ");
        if (seconds > 0) sb.append(seconds).append("s");
        return sb.toString().trim();
    }

    private void initColorMappings() {
        legacyToMiniMessageColors.put("&0", "<black>");
        legacyToMiniMessageColors.put("&1", "<dark_blue>");
        legacyToMiniMessageColors.put("&2", "<dark_green>");
        legacyToMiniMessageColors.put("&3", "<dark_aqua>");
        legacyToMiniMessageColors.put("&4", "<dark_red>");
        legacyToMiniMessageColors.put("&5", "<dark_purple>");
        legacyToMiniMessageColors.put("&6", "<gold>");
        legacyToMiniMessageColors.put("&7", "<gray>");
        legacyToMiniMessageColors.put("&8", "<dark_gray>");
        legacyToMiniMessageColors.put("&9", "<blue>");
        legacyToMiniMessageColors.put("&a", "<green>");
        legacyToMiniMessageColors.put("&b", "<aqua>");
        legacyToMiniMessageColors.put("&c", "<red>");
        legacyToMiniMessageColors.put("&d", "<light_purple>");
        legacyToMiniMessageColors.put("&e", "<yellow>");
        legacyToMiniMessageColors.put("&f", "<white>");
        legacyToMiniMessageColors.put("&l", "<bold>");
        legacyToMiniMessageColors.put("&o", "<italic>");
        legacyToMiniMessageColors.put("&n", "<underlined>");
        legacyToMiniMessageColors.put("&m", "<strikethrough>");
        legacyToMiniMessageColors.put("&k", "<obfuscated>");
        legacyToMiniMessageColors.put("&r", "<reset>");
    }

    @EventHandler(priority = EventPriority.NORMAL)
    public void onChat(AsyncPlayerChatEvent event) {
        String rawMessage = event.getMessage();
        org.bukkit.entity.Player player = event.getPlayer();

        // Mute Check
        de.ayont.lpc.moderation.Mute mute = plugin.getModerationManager().getActiveMute(player.getUniqueId());
        if (mute != null && !player.hasPermission("lpc.mute.bypass")) {
            event.setCancelled(true);
            String timeStr = mute.isPermanent() ? "Permanent" : formatTimeLeft(mute.getExpiry() - System.currentTimeMillis());
            String msgKey = mute.isPermanent() ? "mutes.messages.permanent" : "mutes.messages.muted";
            String muteMsg = plugin.getConfig().getString(msgKey, "<red>You are muted! Reason: {reason}")
                    .replace("{reason}", mute.getReason())
                    .replace("{time}", timeStr);
            plugin.getAdventure().player(player).sendMessage(MiniMessage.miniMessage().deserialize(muteMsg));
            return;
        }

        // Staff Chat
        if (plugin.getConfig().getBoolean("staff-chat.enabled", false)) {
            String prefix = plugin.getConfig().getString("staff-chat.prefix", "#");
            if (rawMessage.startsWith(prefix) && player.hasPermission("lpc.staffchat")) {
                event.setCancelled(true);
                String msg = rawMessage.substring(prefix.length()).trim();
                if (!msg.isEmpty()) {
                    String format = plugin.getConfig().getString("staff-chat.format", "<red>[Staff] {name}: <white>{message}")
                            .replace("{name}", player.getName())
                            .replace("{message}", msg);
                    Component component = MiniMessage.miniMessage().deserialize(format);
                    for (org.bukkit.entity.Player p : org.bukkit.Bukkit.getOnlinePlayers()) {
                        if (p.hasPermission("lpc.staffchat")) {
                            plugin.getAdventure().player(p).sendMessage(component);
                        }
                    }
                    plugin.getAdventure().console().sendMessage(component);
                }
                return;
            }
        }

        // Chat Filter & Moderation
        if (plugin.getConfig().getBoolean("filter.enabled", false) && !player.hasPermission("lpc.filter.bypass")) {
            // Cooldown / Slowmode
            if (!plugin.getModerationManager().canChat(player)) {
                event.setCancelled(true);
                String denyMsg = plugin.getConfig().getString("filter.deny-message", "<red>Please wait {time}s.");
                double timeLeft = plugin.getModerationManager().getTimeLeft(player.getUniqueId());
                plugin.getAdventure().player(player).sendMessage(MiniMessage.miniMessage().deserialize(denyMsg.replace("{time}", String.format("%.1f", timeLeft))));
                return;
            }

            // Caps Limit
            if (plugin.getConfig().getBoolean("filter.caps-limit.enabled", false)) {
                int minLength = plugin.getConfig().getInt("filter.caps-limit.min-length", 5);
                if (rawMessage.length() >= minLength) {
                    double percentage = plugin.getConfig().getDouble("filter.caps-limit.percentage", 70.0);
                    int upperCount = 0;
                    for (char c : rawMessage.toCharArray()) {
                        if (Character.isUpperCase(c)) upperCount++;
                    }
                    if (((double) upperCount / rawMessage.length()) * 100 > percentage) {
                        event.setCancelled(true);
                        plugin.getAdventure().player(player).sendMessage(MiniMessage.miniMessage().deserialize(plugin.getConfig().getString("filter.caps-limit.deny-message", "<red>Too many caps!")));
                        return;
                    }
                }
            }

            // Anti-Discord
            if (plugin.getConfig().getBoolean("filter.anti-discord.enabled", false)) {
                String regex = plugin.getConfig().getString("filter.anti-discord.regex", "discord(?:\\.gg|app\\.com\\/invite|\\.com\\/invite)\\/[a-zA-Z0-9]+");
                if (java.util.regex.Pattern.compile(regex).matcher(rawMessage).find()) {
                    event.setCancelled(true);
                    plugin.getAdventure().player(player).sendMessage(MiniMessage.miniMessage().deserialize(plugin.getConfig().getString("filter.anti-discord.deny-message", "<red>No discord invites!")));
                    return;
                }
            }

            // Anti-Link
            if (plugin.getConfig().getBoolean("filter.anti-link.enabled", false)) {
                String regex = plugin.getConfig().getString("filter.anti-link.regex", "(https?:\\/\\/)?(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)");
                if (java.util.regex.Pattern.compile(regex).matcher(rawMessage).find()) {
                    event.setCancelled(true);
                    plugin.getAdventure().player(player).sendMessage(MiniMessage.miniMessage().deserialize(plugin.getConfig().getString("filter.anti-link.deny-message", "<red>No links!")));
                    return;
                }
            }
        }

        plugin.getModerationManager().updateLastChat(player.getUniqueId());

        // Ignore System
        if (plugin.getConfig().getBoolean("ignore.enabled", false)) {
            event.getRecipients().removeIf(p -> plugin.isIgnored(p.getUniqueId(), player.getUniqueId()));
        }

        // Chat Bubbles
        if (plugin.getConfig().getBoolean("chat-bubbles.enabled", true)) {
            org.bukkit.Bukkit.getScheduler().runTask(plugin, () -> {
                plugin.getChatBubbleManager().spawnBubble(player, rawMessage);
            });
        }
        
        // Chat Channels
        if (plugin.getConfig().getBoolean("channels.enabled", false)) {
             event.setCancelled(true);
             plugin.getChannelManager().sendMessage(player, rawMessage);
             return;
        }

        String message = rawMessage;

        if (event.getPlayer().hasPermission("lpc.chatcolor")) {
            message = message.replaceAll("§", "&");
            for (Map.Entry<String, String> entry : legacyToMiniMessageColors.entrySet()) {
                message = message.replace(entry.getKey(), entry.getValue());
            }
        } else {
            for (Map.Entry<String, String> entry : legacyToMiniMessageColors.entrySet()) {
                message = message.replace(entry.getValue(), entry.getKey());
            }
        }

        if (plugin.getConfig().getBoolean("use-item-placeholder", false) && event.getPlayer().hasPermission("lpc.itemplaceholder")) {
            final ItemStack item = event.getPlayer().getInventory().getItemInMainHand();
            if (!item.getType().equals(Material.AIR)) {
                String itemName = item.getType().toString().toLowerCase().replace("_", " ");
                ItemMeta meta = item.getItemMeta();
                if (meta != null && meta.hasDisplayName()) {
                    try {
                        Component dn = meta.displayName();
                        if (dn != null) {
                            itemName = MiniMessage.miniMessage().serialize(dn);
                        }
                    } catch (NoSuchMethodError e) {
                        itemName = MiniMessage.miniMessage().serialize(
                                LegacyComponentSerializer.builder()
                                        .useUnusualXRepeatedCharacterHexFormat()
                                        .hexColors()
                                        .character('§')
                                        .build()
                                        .deserialize(meta.getDisplayName())
                        );
                    }
                }

                // Use MiniMessage show_item tag which handles NBT/item details automatically
                // Syntax: <hover:show_item:'item_type':count:'nbt'>
                // Since we are using adventure-platform-bukkit, we can also use show_item if supported,
                // but building the tag manually is often safer for legacy Spigot compatibility.
                
                String material = item.getType().getKey().toString();
                String nbt = "";
                try {
                    // Try to get NBT via reflection or Bukkit API if available
                    // For modern Paper/Spigot, item.asHoverEvent() is available via Adventure
                    net.kyori.adventure.text.event.HoverEvent<?> hover = item.asHoverEvent();
                    message = message.replaceFirst("(?i)\\[item]", MiniMessage.miniMessage().serialize(
                            Component.text(itemName).hoverEvent(hover)
                    ));
                } catch (NoClassDefFoundError | NoSuchMethodError e) {
                    // Fallback to simple name if adventure hover fails
                    message = message.replaceFirst("(?i)\\[item]", itemName);
                }
            }
        }

        event.setFormat(LegacyComponentSerializer.builder()
                .useUnusualXRepeatedCharacterHexFormat()
                .hexColors()
                .build()
                .serialize(chatRenderer.render(event.getPlayer(), message, plugin.getAdventure().player(event.getPlayer()))));
    }
}

File: src/main/kotlin/de/ayont/lpc/listener/SpigotChatListener.kt
package de.ayont.lpc.listener

import de.ayont.lpc.LPC
import de.ayont.lpc.renderer.SpigotChatRenderer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import java.util.*
import java.util.regex.Pattern

class SpigotChatListener(private val plugin: LPC) : Listener {
    private val chatRenderer = SpigotChatRenderer(plugin)
    private val legacyToMiniMessageColors = mutableMapOf<String, String>()

    init {
        initColorMappings()
    }

    private fun formatTimeLeft(millis: Long): String {
        if (millis <= 0) return "0s"
        val seconds = millis / 1000 % 60
        val minutes = millis / (60 * 1000) % 60
        val hours = millis / (60 * 60 * 1000) % 24
        val days = millis / (24 * 60 * 60 * 1000)

        val sb = StringBuilder()
        if (days > 0) sb.append(days).append("d ")
        if (hours > 0) sb.append(hours).append("h ")
        if (minutes > 0) sb.append(minutes).append("m ")
        if (seconds > 0) sb.append(seconds).append("s")
        return sb.toString().trim()
    }

    private fun initColorMappings() {
        legacyToMiniMessageColors["&0"] = "<black>"
        legacyToMiniMessageColors["&1"] = "<dark_blue>"
        legacyToMiniMessageColors["&2"] = "<dark_green>"
        legacyToMiniMessageColors["&3"] = "<dark_aqua>"
        legacyToMiniMessageColors["&4"] = "<dark_red>"
        legacyToMiniMessageColors["&5"] = "<dark_purple>"
        legacyToMiniMessageColors["&6"] = "<gold>"
        legacyToMiniMessageColors["&7"] = "<gray>"
        legacyToMiniMessageColors["&8"] = "<dark_gray>"
        legacyToMiniMessageColors["&9"] = "<blue>"
        legacyToMiniMessageColors["&a"] = "<green>"
        legacyToMiniMessageColors["&b"] = "<aqua>"
        legacyToMiniMessageColors["&c"] = "<red>"
        legacyToMiniMessageColors["&d"] = "<light_purple>"
        legacyToMiniMessageColors["&e"] = "<yellow>"
        legacyToMiniMessageColors["&f"] = "<white>"
        legacyToMiniMessageColors["&l"] = "<bold>"
        legacyToMiniMessageColors["&o"] = "<italic>"
        legacyToMiniMessageColors["&n"] = "<underlined>"
        legacyToMiniMessageColors["&m"] = "<strikethrough>"
        legacyToMiniMessageColors["&k"] = "<obfuscated>"
        legacyToMiniMessageColors["&r"] = "<reset>"
    }

    @EventHandler(priority = EventPriority.NORMAL)
    fun onChat(event: AsyncPlayerChatEvent) {
        val rawMessage = event.message
        val player = event.player

        // Mute Check
        val mute = plugin.moderationManager?.getActiveMute(player.uniqueId)
        if (mute != null && !player.hasPermission("lpc.mute.bypass")) {
            event.isCancelled = true
            val timeStr = if (mute.isPermanent) "Permanent" else formatTimeLeft(mute.expiry - System.currentTimeMillis())
            val msgKey = if (mute.isPermanent) "mutes.messages.permanent" else "mutes.messages.muted"
            val muteMsg = plugin.config.getString(msgKey, "<red>You are muted! Reason: {reason}")!!
                .replace("{reason}", mute.reason)
                .replace("{time}", timeStr)
            plugin.adventure?.player(player)?.sendMessage(MiniMessage.miniMessage().deserialize(muteMsg))
            return
        }

        // Staff Chat
        if (plugin.config.getBoolean("staff-chat.enabled", false)) {
            val prefix = plugin.config.getString("staff-chat.prefix", "#")!!
            if (rawMessage.startsWith(prefix) && player.hasPermission("lpc.staffchat")) {
                event.isCancelled = true
                val msg = rawMessage.substring(prefix.length).trim()
                if (msg.isNotEmpty()) {
                    val format = plugin.config.getString("staff-chat.format", "<red>[Staff] {name}: <white>{message}")!!
                        .replace("{name}", player.name)
                        .replace("{message}", msg)
                    val component = MiniMessage.miniMessage().deserialize(format)
                    for (p in Bukkit.getOnlinePlayers()) {
                        if (p.hasPermission("lpc.staffchat")) {
                            plugin.adventure?.player(p)?.sendMessage(component)
                        }
                    }
                    plugin.adventure?.console()?.sendMessage(component)
                }
                return
            }
        }

        // Chat Filter & Moderation
        if (plugin.config.getBoolean("filter.enabled", false) && !player.hasPermission("lpc.filter.bypass")) {
            // Cooldown / Slowmode
            if (plugin.moderationManager?.canChat(player) == false) {
                event.isCancelled = true
                val denyMsg = plugin.config.getString("filter.deny-message", "<red>Please wait {time}s.")!!
                val timeLeft = plugin.moderationManager?.getTimeLeft(player.uniqueId) ?: 0.0
                plugin.adventure?.player(player)?.sendMessage(MiniMessage.miniMessage().deserialize(denyMsg.replace("{time}", String.format("%.1f", timeLeft))))
                return
            }

            // Caps Limit
            if (plugin.config.getBoolean("filter.caps-limit.enabled", false)) {
                val minLength = plugin.config.getInt("filter.caps-limit.min-length", 5)
                if (rawMessage.length >= minLength) {
                    val percentage = plugin.config.getDouble("filter.caps-limit.percentage", 70.0)
                    var upperCount = 0
                    for (c in rawMessage.toCharArray()) {
                        if (Character.isUpperCase(c)) upperCount++
                    }
                    if ((upperCount.toDouble() / rawMessage.length) * 100 > percentage) {
                        event.isCancelled = true
                        plugin.adventure?.player(player)?.sendMessage(MiniMessage.miniMessage().deserialize(plugin.config.getString("filter.caps-limit.deny-message", "<red>Too many caps!")!!))
                        return
                    }
                }
            }

            // Anti-Discord
            if (plugin.config.getBoolean("filter.anti-discord.enabled", false)) {
                val regex = plugin.config.getString("filter.anti-discord.regex", "discord(?:\\.gg|app\\.com\\/invite|\\.com\\/invite)\\/[a-zA-Z0-9]+")!!
                if (Pattern.compile(regex).matcher(rawMessage).find()) {
                    event.isCancelled = true
                    plugin.adventure?.player(player)?.sendMessage(MiniMessage.miniMessage().deserialize(plugin.config.getString("filter.anti-discord.deny-message", "<red>No discord invites!")!!))
                    return
                }
            }

            // Anti-Link
            if (plugin.config.getBoolean("filter.anti-link.enabled", false)) {
                val regex = plugin.config.getString("filter.anti-link.regex", "(https?:\\/\\/)?(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)")!!
                if (Pattern.compile(regex).matcher(rawMessage).find()) {
                    event.isCancelled = true
                    plugin.adventure?.player(player)?.sendMessage(MiniMessage.miniMessage().deserialize(plugin.config.getString("filter.anti-link.deny-message", "<red>No links!")!!))
                    return
                }
            }
        }

        plugin.moderationManager?.updateLastChat(player.uniqueId)

        // Ignore System
        if (plugin.config.getBoolean("ignore.enabled", false)) {
            event.recipients.removeIf { p -> plugin.playerSettingsManager.isIgnored(p.uniqueId, player.uniqueId) }
        }

        // Chat Bubbles
        if (plugin.config.getBoolean("chat-bubbles.enabled", true)) {
            Bukkit.getScheduler().runTask(plugin, Runnable {
                plugin.chatBubbleManager?.spawnBubble(player, rawMessage)
            })
        }
        
        // Chat Channels
        if (plugin.config.getBoolean("channels.enabled", false)) {
             event.isCancelled = true
             plugin.channelManager?.sendMessage(player, rawMessage)
             return
        }

        var message = rawMessage

        if (event.player.hasPermission("lpc.chatcolor")) {
            message = message.replace("§", "&")
            for ((key, value) in legacyToMiniMessageColors) {
                message = message.replace(key, value)
            }
        } else {
            for ((key, value) in legacyToMiniMessageColors) {
                message = message.replace(value, key)
            }
        }

        if (plugin.config.getBoolean("use-item-placeholder", false) && event.player.hasPermission("lpc.itemplaceholder")) {
            val item = event.player.inventory.itemInMainHand
            if (item.type != Material.AIR) {
                var itemName = item.type.toString().lowercase(Locale.getDefault()).replace("_", " ")
                val meta = item.itemMeta
                if (meta != null && meta.hasDisplayName()) {
                    try {
                        val dn = meta.displayName()
                        if (dn != null) {
                            itemName = MiniMessage.miniMessage().serialize(dn)
                        }
                    } catch (e: NoSuchMethodError) {
                        itemName = MiniMessage.miniMessage().serialize(
                            LegacyComponentSerializer.builder()
                                .useUnusualXRepeatedCharacterHexFormat()
                                .hexColors()
                                .character('§')
                                .build()
                                .deserialize(meta.displayName)
                        )
                    }
                }
                
                try {
                    val hover = item.asHoverEvent()
                    message = message.replaceFirst("(?i)\\[item]".toRegex(), MiniMessage.miniMessage().serialize(
                        Component.text(itemName).hoverEvent(hover)
                    ))
                } catch (e: Throwable) {
                    message = message.replaceFirst("(?i)\\[item]".toRegex(), itemName)
                }
            }
        }

        val adventurePlayer = plugin.adventure?.player(event.player)
        if (adventurePlayer != null) {
            event.format = LegacyComponentSerializer.builder()
                .useUnusualXRepeatedCharacterHexFormat()
                .hexColors()
                .build()
                .serialize(chatRenderer.render(event.player, message, adventurePlayer))
        }
    }
}
